ProjFS.Mac/prjfs-tracectl/prjfs-tracectl.m

```diff
@@ -0,0 +1,154 @@
+                fprintf(stderr, "--path-filter: Currently only one path filter is supported\n");
+                goto CleanupAndFail;
+            }
+            
+            pathFilterString = CFStringCreateWithCString(kCFAllocatorDefault, optarg, kCFStringEncodingUTF8);
+            
+            break;
+        case 0:
+            printf("Processing argument %s\n", argv[optind]);
+            break;
+        default:
+            fprintf(stderr, "TODO: %u\n", ch);
+            goto CleanupAndFail;
+        }
+    }
+    
+    if ((tracingEnabled ^ tracingDisabled) == 0)
+    {
+        fprintf(stderr, "Must use exactly one of --enable or --disable");
+        goto CleanupAndFail;
+    }
+    else if (pathFilterString == NULL)
+    {
+        fprintf(stderr, "--path-filter is required");
+        goto CleanupAndFail;
+    }
+
+    if (tracingDisabled)
+    {
+        if (pathFilterString != NULL)
+        {
+            CFRelease(pathFilterString);
+        }
+        return kCFBooleanFalse;
+    }
+    else
+    {
+        NSMutableDictionary* settings = [[NSMutableDictionary alloc] initWithCapacity:5];
+        if (traceAllVnodeEvents)
+        {
+            [settings setObject:[NSNumber numberWithBool:YES] forKey:@"vnode-events-all"];
+        }
+        else
+        {
+            [settings setObject:[NSNumber numberWithBool:(traceDeniedVnodeEvents != 0)] forKey:@"vnode-events-denied"];
+            [settings setObject:[NSNumber numberWithBool:(traceProviderMessagingEvents != 0)] forKey:@"vnode-message-events"];
+        }
+        
+        if (pathFilterString != NULL)
+        {
+            [settings setObject:(__bridge_transfer NSString*)pathFilterString forKey:@"path-filter"];
+        }
+        
+        if (useVnodeActionFilter)
+        {
+            [settings setObject:@(vnodeActionFilter) forKey:@"vnode-action-filter-mask"];
+        }
+        
+        [settings setObject:[NSNumber numberWithBool:traceAllFileopEvents] forKey:@"fileop-all-events"];
+        
+        NSLog(@"Settings dictionary: %@", settings);
+        
+        return (__bridge_retained CFDictionaryRef)settings;
+    }
+    
+CleanupAndFail:
+    if (pathFilterString != NULL)
+    {
+        CFRelease(pathFilterString);
+    }
+    
+    return NULL;
+}
+
+int main(int argc, char* argv[])
+{
+    CFTypeRef traceSettings = GenerateEventTracingSettings(argc, argv);
+    if (traceSettings == NULL)
+    {
+        fprintf(stderr, "Failed to generate tracing settings\n");
+        return 1;
+    }
+
+    CFDictionaryRef matchDict = IOServiceMatching(PrjFSServiceClass);
+    io_service_t prjfsService = IOServiceGetMatchingService(kIOMasterPortDefault, matchDict); // matchDict consumed
+    
+    if (prjfsService == IO_OBJECT_NULL)
+    {
+        CFRelease(traceSettings);
+        fprintf(stderr, "PrjFS Service object not found.\n");
+        return 1;
+    }
+    
+    IORegistryEntrySetCFProperty(prjfsService, CFSTR(PrjFSEventTracingKey), traceSettings);
```

Comment:
Sounds sensible! Detecting process death will require switching to a user client based approach, so there's a persistent connection between the kext and the trace process which gets severed when the process dies. I'll try to get that done in the next few days.

Reply:
While implementing this it occurred to me that if we're going to go to the effort of an extra user client type, we could push the trace events directly to that client rather than going via the KextLog daemon. This has a few advantages:

 * The trace logs are *right there* on stdout from the `prjfs-tracectl` command. No need to hunt around in Console or use `log stream`, and no issues with which type of build is running.
 * We can make the trace buffer a *lot* larger than the KextLog buffer, so we're less likely to drop events. We can even make the buffer size configurable.

I'd probably use the same message framing protocol as with KextLog as that means we can share the encoding/decoding code, which could be extra handy when we add support for logging stack traces in KextLog.

What do you think of this approach @jeschu1? It's a little bit of extra work but nothing massive.